core/src/sumeragi/main_loop.rs

```diff
@@ -77,6 +77,9 @@ where
     pub message_sender: Mutex<mpsc::SyncSender<MessagePacket>>,
     /// Receiver channel.
     pub message_receiver: Mutex<mpsc::Receiver<MessagePacket>>,
+    /// Only used in testing. Causes the genesis peer to withhold blocks when it
+    /// is the proxy tail.
+    pub debug_force_soft_fork: bool,
```

Comment:
I wonder could we test soft-forks by other means because i'm not really fan of such approach. 

For example we can construct 2 `BlockSync` update message (with soft-fork in them) and feed them to sumeragi.

Reply:
it's also possible to use compilation flags

Reply:
I will find a different solution.

Reply:
While compile-time forcing is going to lead to quick testing, it doesn't seem to be the kind of thing that can be covered with unit tests. Integration tests, on the other hand need to work on the standard binary (because conditional compilation itself can produce spurious failures/successes). 


Can we come up with a realistic scenario that can come up in a deployment, force a soft-fork programmatically and externally (e.g. in the bash-script tests that we already have), and use that to test in the distributed environment?

Reply:
> Can we come up with a realistic scenario that can occur in a deployment, force a soft-fork programmatically and externally (e.g. in the bash-script tests that we already have), and use that to test in the distributed environment?

Conclusion was no. It must be triggered from the inside in order for it to happen consistently.